compiler/optimizer/LoopVersioner.cpp

```diff
@@ -1276,6 +1276,261 @@ bool TR_LoopVersioner::detectInvariantCheckCasts(List<TR::TreeTop> *trees)
    return foundInvariantTrees;
    }
 
+/**
+ * \brief Determine whether it is possible to version \p ifNode w.r.t. an
+ * extremum of a non-invariant function of an IV.
+ *
+ * \param[in]  ifNode The conditional node to analyze
+ * \param[out] out    The results of analyzing the conditional node, if
+ *                    versioning is possible.
+ *
+ * \return true when versioning is possible, false otherwise.
+ */
+bool TR_LoopVersioner::isVersionableIfWithExtremum(
+   TR::TreeTop *ifTree, ExtremumConditional *out)
+   {
+   TR::Node *ifNode = ifTree->getNode();
+   ExtremumConditional excond = {};
+   excond.varyingChildIndex = -1; // invalid
+   *out = excond;
+
+   bool withMin = ifNode->isVersionableIfWithMinExpr();
+   bool withMax = ifNode->isVersionableIfWithMaxExpr();
+   if (!withMin && !withMax)
+      return false;
+
+   TR_ASSERT_FATAL_WITH_NODE(
+      ifNode, !withMin || !withMax, "both min and max flags are set");
+
+   static const bool enable = feGetEnv("TR_dontVersionIfWithExtremum") == NULL;
+   if (!enable)
+      return false;
+
+   TR_ASSERT_FATAL_WITH_NODE(
+      ifNode, ifNode->getNumChildren() == 2, "unexpected number of children");
+
+   if (!ifNode->getChild(0)->getDataType().isIntegral()
+       || ifNode->getChild(0)->getDataType() != ifNode->getChild(1)->getDataType())
```

Comment:
When can the `ifNode` have children that are not of the same datatype ? Maybe this is guarding against bad IL (?)

Reply:
Yeah, I suppose that would only happen in bad IL. It doesn't necessarily make much sense to check it here, given that below I implicitly assume that all encountered arithmetic ops are well-typed